horovod/_keras/__init__.py

```diff
@@ -41,34 +53,129 @@ def get_gradients(self, loss, params):
             In DistributedOptimizer, get_gradients() is overriden to also
             allreduce the gradients before returning them.
             """
-            self._get_gradients_used = True
-            gradients = super(self.__class__, self).get_gradients(loss, params)
-            if hvd.size() > 1:
-                averaged_gradients = []
-                with tf.name_scope(self._name + "_Allreduce"):
-                    for grad in gradients:
-                        if grad is not None:
+            def init_aggregation_vars():
+                v = tf.get_collection('aggregation_variables')
+                vars_init_op = tf.variables_initializer(v)
+                sess = tf.keras.backend.get_session(op_input_list=())
+
+                with tf.variable_scope("aggregation_variables"):
+                    self.counter = tf.get_variable(
+                        "aggregation_counter", shape=(), dtype=tf.int32,
+                        trainable=False, initializer=tf.zeros_initializer())
+                    if self._aggregation_frequency > 1:
+                        for idx, grad in enumerate(self.grads):
                             if self._sparse_as_dense and \
                                     isinstance(grad, tf.IndexedSlices):
                                 grad = tf.convert_to_tensor(grad)
+                            grad_aggregation_variable_name = str(idx)
```

Comment:
Should we put a check in here that `grad` is not `isinstance(grad, tf.IndexedSlices)`? Haven't dove too deep on this but it seems to me that our current gradient aggregation techniques would not support it.

Reply:
just fail in that case?

Reply:
Yeah just print  if `aggregation_frequency > 1` we don't support sparse communication. On second look we might need to do this before the `sparse_as_dense` check also.

Reply:
Can we move this up above ` if self._sparse_as_dense and...`